modules/matterhorn-index-service/src/main/java/org/opencastproject/index/service/impl/index/event/EventIndexUtils.java

```diff
@@ -640,7 +640,7 @@ public static void updateComments(String eventId, boolean hasComments, boolean h
     if (!hasOpenComments && needsCutting)
       throw new IllegalStateException(
               "Invalid comment update request: You can't have an needs cutting comment without having any open comments!");
-    Event event = getEvent(eventId, organization, user, searchIndex);
+    Event event = getOrCreateEvent(eventId, organization, user, searchIndex);
```

Comment:
The name _getOrCreateEvent_ suggests that an event would be created if not already existing when updating a comment... that seems wrong...

Some words on this?

Reply:
All comments must have associated events. If you have comments without events, your database is inconsistent.

The index rebuild clears the whole index and recreates the data from each service asynchronously. Each of these services (asset manager, scheduler, workflow,...) will provide the data for the update. As it should run asynchronous, the update messages should be handled out of order. I've tested this behavior on my local workstation with <10 events. The event-comment update messages was handled before the messages from the asset manager (in most cases). Therefore there wasn't an event entry in the index yet. As a result the comment data was lost.

Reply:
I don't think that the mere existence of a second class citizen (comment) should bring a first class citizen (event) into existence. With other words: If the event has been deleted, the existence of a comment for this event should not cause the event to be revived even though this would be database inconsistency. Otherwise, we start to revert user actions which is very undesirable. In particular, because there is no obvious way to see which events exist just because of inconsistent comments.

Are you sure that the indices are created concurrently? The log very much suggests that the indices of services are build one after another... 

Please also note that there was a bug since 2.x that caused comments not to be deleted, so this change actually would revive a large number of events in almost any Opencast instance (revived 700+ events on one single  tenant on our staging environment, for example).

Reply:
I agree with that. Possible database inconsistance should be handled in the migration script.